### PR TITLE
Rewrite dracut zfs module to fix #3875 and #3876 and other issues

### LINKED ISSUE #3875: 0.6.5 Regression: Exporting the pool on shutdown in dracut dirties vdev labels and pollutes zpool history.
The new dracut code that does pool export on shutdown in dracut dirties the vdev labels, which breaks `zpool import -T` and `zpool import -F` because the kernel cannot distinguish between a vdev configuration change and the export. It also pollutes the `zpool history` with commands that the user never actually ran on the pool.

I did not pay too much attention to what was happening in the dracut module because people on Gentoo use genkernel and people at work do not worry about / on ZFS, but I am starting to pay attention because the incorrect behavior could make catastrophic failures that could be recovered with some minor assistance from the ZoL project irrecoverable without major assistance.

I tried export on shutdown in 2012 in Gentoo and abandoned it after discovering the problems it caused:

https://sources.gentoo.org/cgi-bin/viewvc.cgi/gentoo-x86/sys-fs/zfs/files/zfs-shutdown?view=log&pathrev=MAIN

The export on shutdown behavior in the dracut zfs module should be removed.

### PR DESCRIPTION
**Pull request by ryao**
I have tested this against dracut 43 on my workstation. It boots my system just like genkernel's ZFS branch and fixes several issues:

* No more export of pools on shutdown dirtying the vdev labels or polluting the history file (regression fix)
* No more coherency issues between the initramfs archive and system in the hostid, zpool.cache and vdev.conf files.
* No more failed boots from faulted pools when a vdev is added to the root pool.
* The unnecessary awk command is no longer included. Similarly, the head and hostid utilities have become unnecessary and are no longer included as well.

There are a few behavior changes:

* We load from the cachefile rather than importing. If an import is required due to an absent cachefile or pool absent from the cachefile, the system will drop to a shell and the user is expected to do it. Instructions are provided in this situation.
* zfs:AUTO support has been dropped. This was ambiguous in two ways. One is that `zfs:AUTO` should mean to import a pool named AUTO and mount the root dataset as /. Two is that the behavior is non-deterministic when multiple pools have the bootfs. Users of `zfs:AUTO` will receive the instructions for a missing cachefile while users of autodetection through `zfs` and `zfs:` will receive explicit instructions about putting the dataset name on the commandline. Few, if any users, will be affected by this given that I am continually unable to find users of this functionality. The vast majority of users use GRUB2, which generates the commandline for users with `root=ZFS=$BOOTFS`.

I need to test against the versions of dracut used in RHEL6 and RHEL7. I intend to bring these changes into Gentoo for the few dracut users there afterward. I also need to improve the commit messages. I opened the pull request in advance of that for feedback.

### CODE REVIEW

File: contrib/dracut/90zfs/mount-zfs.sh.in [hunk 1]
```diff
@@ -13,32 +13,69 @@ esac
...
+then
+	if ! mount_dataset "${ZFS_DATASET}" ; then
+		/sbin/zpool export -F "${ZFS_POOL}"
+		warn "Failed to mount \"${ZFS_DATASET}\"."
+		rootok=0
+		return 1
+	fi
+
+	# Copy important files to /etc/zfs
+	for i in zpool.cache vdev.conf
```
Comment (ryao) [inline contrib/dracut/90zfs/mount-zfs.sh.in]: As a note to myself, a `udevadm trigger && udevadm settle` is needed when vdev.conf is read.

**Comment by evolarium** [comment]
As I mentioned in #3605, I don't believe the current dracut scripts work when systemd >=220 is used.

[Related bug with dmsquash and systemd 220.](https://bugs.freedesktop.org/show_bug.cgi?id=90913)

systemd-fstab-generator used to fail gracefully, allowing the dracut-mount hook to mount the zfs root.  However, it now will generate a faulty sysroot.mount, stopping the boot before the dracut-mount hook.  I think a zfs-generator has to be written to generate the sysroot.mount, rather than doing it in the dracut-mount hook, [as was done with the dmsquash.](http://git.kernel.org/cgit/boot/dracut/dracut.git/commit/?id=2c66c3b)

RHEL 7 uses an older version of systemd, so it won't be an issue there.

**Comment by prometheanfire** [comment]
recheck

**Comment by Rudd-O** [comment]
@evolarium  I have fixed the systemd >= 220 issue, take a look at my pull request in this repo.

@ryao  I would love to see how these patches of yours look, when rebased on top of that pull request.

**Comment by behlendorf** [comment]
@Rudd-O's work has been merged to master.  Let's get any additional dracut fixes rebased.

d402c18 A collection of dracut fixes

**Comment by Rudd-O** [comment]
I am happy to test with my media center, which is my most complex setup, including but not limited to LUKS below ZFS.

**Comment by Rudd-O** [comment]
@ryao can you test your changes under Fedora with the following process?

1. Build RPMs of your code (SPL and ZFS).
2. Place results in directory `.../path/to/rpms`.
3. Clone https://github.com/Rudd-O/zfs-fedora-installer to a local folder.
4. In that folder, run `sudo ./install-fedora-on-zfs --use-prebuilt-rpms=.../path/to/rpms/ --root-password=password --chown=$USER --chgrp=$USER root.img`.
5. Boot the `root.img` image using QEMU now; you can find information on how to do that online. You want to use qemu-system-x86_64 in accel=kvm mode or else it will be horribly slow.

This will tell us whether your changes work properly against Fedora.

Also, is there any chance you could help us by splicing the different conceptual changes into separate patches against HEAD?

Thank you in advance.

**Comment by behlendorf** [comment]
Closing as out of date.  Commit 7e8a2d0b755ea67291d3f576069706e43abc5e31 updates dracut to pass the `-F` flag on shutdown to `zpool export` to prevent  it from dirtying the  labels.  We can purse the other ideas proposed here in new PRs.

**Pull request closed without merging**